scipy/interpolate/_bspl.pyx

```diff
@@ -473,3 +475,193 @@ def _make_design_matrix(const double[::1] x,
+            Output values of the b-spline at given ``xi`` points.
+
+        Notes
+        -----
+
+        This function is essentially equivalent to the following: given an
+        N-dimensional vector ``x = (x1, x2, ..., xN)``, iterate over the
+        dimensions, form linear combinations of products,
+        B(x1) * B(x2) * ... B(xN) of (k+1)**N b-splines which are non-zero
+        at ``x``. 
+
+        Since b-splines are localized, the sum has (k+1)**N non-zero elements.
+
+        If ``i = (i1, i2, ..., iN)`` is a vector if intervals of the knot
+        vectors, ``t[d, id] <= xd < t[d, id+1]``, for ``d=1, 2, ..., N``, then
+        the core loop of this function is nothing but
+
+        ```
+        result = 0
+        iters = [range(i[d] - self.k[d], i[d] + 1) for d in range(ndim)]
+        for idx in itertools.product(*iters):
+            term = self.c[idx] * np.prod([B(x[d], self.k[d], idx[d], self.t[d])
+                                          for d in range(ndim)])
+            result += term
+        ```
+
+        For efficiency reasons, we iterate over the flattened versions of the
+        arrays.
+
+        """
+        cdef:
+            npy_intp ndim = len(t)
+
+            # 'intervals': indices for a point in xi into the knot arrays t
+            npy_intp[::1] i = np.empty(ndim, dtype=int)
+
+            # container for non-zero b-splines at each point in xi
+            double[:, ::1] b = np.empty((ndim, max(k) + 1), dtype=float)
+
+            const double[::1] xv     # an ndim-dimensional input point
+            double xd               # d-th component of x
+
+            const double[::1] td    # knots in dimension d
+
+            npy_intp kd             # d-th component of k
+
+            npy_intp i_c      # index to loop over range(num_c_tr)
+            npy_intp iflat    # index to loop over (k+1)**ndim non-zero terms
+            npy_intp volume   # the number of non-zero terms
+            const npy_intp[:] idx_b   # ndim-dimensional index corresponding to iflat
+
+            int out_of_bounds
+            npy_intp idx_cflat_base, idx
+            double factor
+            double[::1] wrk = np.empty(2*max(k) + 2, dtype=float)
+
+        if xi.shape[1] != ndim:
+            raise ValueError(f"Expacted data points in {ndim}-D space, got"
+                             f" {xi.shape[1]}-D points.")
+
+        if out.shape[0] != xi.shape[0]:
+            raise ValueError(f"out and xi are inconsistent: cxpected"
+                             f" {xi.shape[0]} output values, got"
+                             f" {out.shape[0]}.")
+        if out.shape[1] != num_c_tr:
+            raise ValueError(f"out and c are inconsistent: num_c={num_c_tr} "
+                             f" and out.shape[1] = {out.shape[1]}.")
+
+        # the number of non-zero terms for each point in ``xi``.
+        volume = 1
+        for d in range(ndim):
+            volume *= k[d] + 1
```

Comment:
Can this loop be included in the `nogil` section?

Reply:
Yup, moved.